### PR TITLE
[fix](type) Data precision is lost when converting DOUBLE type data to DECIMAL (#17191)

### LINKED ISSUE #17191: [Bug] Data precision is lost when converting DBOUBLE type data to DECIMAL
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

1.2.2

### What's Wrong?

Data precision is lost when using the CAST function to convert DOUBLE type to DECIMAL type. 133059.33 will be converted to 133059.32999999.
image

image
image

### What You Expected?

When using the CAST function to convert 133059.33 of the DOUBLE type to the DECIMAL type, it is still 133059.33

### How to Reproduce?

_No response_

### Anything Else?

SQL：

SELECT
	CAST(ROUND(CAST(t_a_0.`C_8e43b548775420c62858337ce2152b8e` AS DECIMAL(10,8)), 8)  AS DECIMAL(10,8)) castRoundCast,
	ROUND(CAST(t_a_0.`C_8e43b548775420c62858337ce2152b8e` AS DECIMAL(10,8)), 8) roundCast,
	CAST(t_a_0.`C_8e43b548775420c62858337ce2152b8e` AS DECIMAL(10,8)) castDecimal,
	CAST(t_a_0.`C_8e43b548775420c62858337ce2152b8e` AS double) castDouble
FROM
	`ds_f55411c9_0615_47dc_b9bb_1f2f055049b3` t_a_0 
WHERE
	t_a_0.`C_8e43b548775420c62858337ce2152b8e` = '133059.33';

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by herry2038**
# Proposed changes

Issue Number: close #17191

## Problem summary

1. Decimal template class 's construction method  using strict convernion from double to int ，it lose data precision , using round instead .
2. DecimalV2 's value 's scale is always 9, so when output to mysqlwriter, the to string function will lose data precision too, fix it .
3. Add regression test regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy for the problem.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

### CODE REVIEW

File: be/src/runtime/decimalv2_value.cpp [hunk 1]
```diff
@@ -373,7 +373,19 @@ std::string DecimalV2Value::to_string(int scale) const {
             }
         }
     } else {
-        frac_val = frac_val / SCALE_TRIM_ARRAY[scale];
+        // frac_val = frac_val / SCALE_TRIM_ARRAY[scale];
```
Comment (Yukang-Lian) [inline be/src/runtime/decimalv2_value.cpp:376]: please remove useless comment.

File: be/src/vec/core/types.h [hunk 2]
```diff
@@ -308,9 +308,20 @@ struct Decimal {
     DECLARE_NUMERIC_CTOR(Int64)
     DECLARE_NUMERIC_CTOR(UInt32)
     DECLARE_NUMERIC_CTOR(UInt64)
-    DECLARE_NUMERIC_CTOR(Float32)
-    DECLARE_NUMERIC_CTOR(Float64)
+    // FIX 17191
+    //DECLARE_NUMERIC_CTOR(Float32)
+    //DECLARE_NUMERIC_CTOR(Float64)
```
Comment (Yukang-Lian) [inline be/src/vec/core/types.h:313]: ditto

File: regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_cast_double_to_decimal") {
+    def db = "test_double"
+    sql "CREATE DATABASE IF NOT EXISTS ${db}"
+    sql "use ${db}"
```
Comment (Yukang-Lian) [inline regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:21]: We don't need to use these 3 lines. Actually, regression test framework will create db automaticlly. You just need to create table.
```suggestion
     def tableName = "tbl_test_xxx"
    sql """DROP TABLE IF EXISTS ${tableName}"""
    sql """
            CREATE TABLE IF NOT EXISTS ${tableName} (xxx)
            """
           
```

File: regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	sql "insert into test5 values(5,15.9999)"
+	sql "insert into test5 values(6,0.9999)"
+	sql "insert into test5 values(7,15.0245)"
+	sql "insert into test5 values(11,-15.025)"
+	sql "insert into test5 values(12,-15.024999999999)"
+	sql "insert into test5 values(13,-15.33)"
+	sql "insert into test5 values(14,-15.999999995)"
+	sql "insert into test5 values(15,-15.9999)"
+	sql "insert into test5 values(16,-0.9999)"
+	sql "insert into test5 values(17,-15.0245)"
```
Comment (Yukang-Lian) [inline regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:37]: maybe you can use one line `insert` statement.

File: regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		logger.info("convert double to decimal test key: ${keyAndExcpetedValue[0]} -> ${keyAndExcpetedValue[1]} ${keyAndExcpetedValue[2]} ${keyAndExcpetedValue[3]} ${keyAndExcpetedValue[4]}")
+		def s = "select cast(d as decimal(10,3)), cast(d as decimalv3(10,3)), cast(d as decimal(10,8)), cast(d as decimalv3(10,8)) from test5 where id=" + keyAndExcpetedValue[0]
+		def result = sql s
+		logger.info("sql result: ${result[0][0]} ${result[0][1]} ${result[0][2]} ${result[0][3]}")
+
+		for (int i=0 ;i<4 ;i++) {
+			//assertEquals(keyAndExcpetedValue[i+1], result[0][i].toString())
+			assertEquals(keyAndExcpetedValue[i+1], result[0][i].toString())
+		}
+	}
```
Comment (Yukang-Lian) [inline regression-test/suites/cast_double_to_decimal/test_cast_double_to_decimal.groovy:67]: we don't need to use `assertequals`, use `qt_select` statement and generate `.out` file.
reference https://doris.incubator.apache.org/zh-CN/community/developer-guide/regression-testing

File: be/src/vec/core/types.h (hunk 2 above)
Comment (herry2038) [reply be/src/vec/core/types.h:314]: Thanks, done!

File: be/src/runtime/decimalv2_value.cpp (hunk 1 above)
Comment (herry2038) [reply be/src/runtime/decimalv2_value.cpp]: Done

**Comment by herry2038** [comment]
@Yukang-Lian 帮我再看看，谢谢！

**Review by Yukang-Lian — DISMISSED** [review]
Good job! LGTM

**Comment by Yukang-Lian** [comment]
@Gabriel39 PTAL

**Review by Gabriel39 — DISMISSED** [review]
LGTM

**Comment by Gabriel39** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Comment by Gabriel39** [comment]
Merged. Thanks for your contribution!

**Pull request merged**